relations/jf/src/merge.rs

```diff
@@ -0,0 +1,378 @@
+
+        let second_old_note = compute_note(
+            token_id,
+            second_old_token_amount,
+            second_old_trapdoor,
+            second_old_nullifier,
+        );
+
+        let new_token_amount = first_old_token_amount + second_old_token_amount;
+        let new_trapdoor = [5; 4];
+        let new_nullifier = [6; 4];
+
+        let new_note = compute_note(token_id, new_token_amount, new_trapdoor, new_nullifier);
+
+        let first_leaf_index = 0u64;
+        let first_uid = BigUint::from(first_leaf_index);
+        let first_value = convert_array(first_old_note);
+
+        let second_leaf_index = 1u64;
+        let second_uid = BigUint::from(second_leaf_index);
+        let second_value = convert_array(second_old_note);
+
+        let tree = RescueSparseMerkleTree::from_kv_set(
+            MERKLE_TREE_HEIGHT,
+            &[
+                (first_uid.clone(), first_value),
+                (second_uid.clone(), second_value),
+            ],
+        )
+        .expect("create Merkle tree from k-v pairs");
+
+        let (first_value_retrieved, first_merkle_proof) = tree
+            .lookup(&first_uid)
+            .expect_ok()
+            .expect("lookup first old note in Merkle tree");
+
+        assert_eq!(first_value, first_value_retrieved);
+        assert!(tree
+            .verify(&first_uid, first_merkle_proof.clone())
+            .expect("membership verified"));
+
+        let (second_value_retrieved, second_merkle_proof) = tree
+            .lookup(&second_uid)
+            .expect_ok()
+            .expect("lookup second old note in Merkle tree");
+
+        assert_eq!(second_value, second_value_retrieved);
+        assert!(tree
+            .verify(&second_uid, second_merkle_proof.clone())
+            .expect("membership verified"));
+
+        let merkle_root = tree.commitment().digest().into_bigint().0;
+
+        let public = MergePublicInput {
+            token_id,
+            first_old_nullifier,
+            second_old_nullifier,
+            new_note,
+            merkle_root,
+        };
+
+        let private = MergePrivateInput {
+            first_old_trapdoor,
+            second_old_trapdoor,
+            new_trapdoor,
+            new_nullifier,
+            first_merkle_path: first_merkle_proof,
+            second_merkle_path: second_merkle_proof,
+            first_leaf_index,
+            second_leaf_index,
+            first_old_note,
+            second_old_note,
+            first_old_token_amount,
+            second_old_token_amount,
+            new_token_amount,
+        };
+
+        MergeRelation::new(public, private)
+    }
+
+    fn merge_constraints_correctness(relation: MergeRelation) -> bool {
+        let circuit = MergeRelation::generate_circuit(&relation).unwrap();
+
+        match circuit.check_circuit_satisfiability(&relation.public_input()) {
+            Ok(_) => true,
+            Err(why) => {
+                println!("circuit not satisfied: {}", why);
```

Comment:
I'd return the reason here from the method : `anyhow::Result` and use the `why` as the _reason_. Then, on the call-site you can unwrap or `?` and see it in the logs by default, without having to rerun with `-- --nocapture`.

Reply:
yet sometimes a failure is the desired outcome. in that case assertion panics and you get println! statements only when some tests failed

Reply:
You can still do that. If you expect `Ok` just `assert!(res.is_ok())` or `assert!(res.is_err())` otherwise. You'll get the error statement if the test fails as well b/c test runner prints the stack (reason).

Reply:
then it's just a different way of writing the same behavior, hardly an improvement?

Reply:
I think my version is just cleaner - bubbles up the error to the caller instead of hiding it and relying on some assumptions of the test framework.

It's a nit nevertheless so you'll decide.